File: FootballStats.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <iomanip>
#include <math.h>  

using namespace std;

//----------------------------Declarations-------------------------------------------------------------------

double calcMean(vector <double>& vec),calcStdDeviation(vector <double>& dev,double mean);
double calcMax(vector <double>& max),calcMin(vector <double>& min); 
void readFile(),parseIt(),calcPSM(),printTeamStats();
double rTest();
vector<string> TeamName, TeamData;
vector<double> asM,py,dpy,ry,dry,to,dto,pendif,rettd,psm;

//---------------------------Main function-------------------------------------------------------------------

int main(int argc, char *argv[]) {     

    readFile();        
    parseIt();
    calcPSM();         
    rTest();           
    printTeamStats();  

    return 0;
}

//-------------This function will read text file "teamStats.txt"---------------------------------------------

void readFile(){                        
    string singleLine = "";
    ifstream infile;
    infile.open("teamStats.txt");

   
    while (!infile.eof()){
        getline(infile, singleLine);
        size_t key = singleLine.find("|");                //make your delimter = " | "
        string name = singleLine.substr (0,key-1);        //split string delimiter is found and make it team Name
        string data = singleLine.substr (key+1);          //make rest string as team data.
        TeamName.push_back(name);                         //store all first into teamName vector.
        TeamData.push_back(data);                         //store all second into teamData vector.
    }

    infile.close();                                      //closing the file is necessary.
}

//-------------This function will parse individual read(stored) line into teamName and teamData.-------------

void parseIt(){
       for(int i = 0; i<TeamData.size();i++){
        istringstream iss(TeamData[i]);

        int index = 0;

        while(!iss.eof()){
           string parsedString;
           iss >> parsedString;                  //parse teamData string into pieces to get all statstical values
           double convertedToDouble = stod(parsedString);     //convert all parsed string into double.

            if(index == 0)
                asM.push_back(convertedToDouble);
            else if(index == 1)
                py.push_back(convertedToDouble);
            else if(index == 2)
                dpy.push_back(convertedToDouble);
            else if(index== 3)
                ry.push_back(convertedToDouble);
            else if(index== 4)
                dry.push_back(convertedToDouble);
            else if(index== 5)
                to.push_back(convertedToDouble);
            else if(index== 6)
                dto.push_back(convertedToDouble);
            else if(index== 7)
                pendif.push_back(convertedToDouble);
            else if(index== 8)
                rettd.push_back(convertedToDouble);
            else 
                cout << "Error in Storing Data !\n";

            index++;
        }
    }
}

//-------------This function will calculate PSM of all the teams based on given stats.-----------------------

void calcPSM(){
    for(int i = 0; i < TeamName.size(); i++){
        double ans = 3.17 * (rettd[i]) - 0.06 * (pendif[i]) + 61.67 * (py[i]) + 26.44 * (ry[i]) - 2.77 * (to[i]) - 67.5 * (dpy[i]) - 22.79 * (dry[i]) + 3.49 * (dto[i]);
        psm.push_back(ans);
    }
}

//--------------This function will calculate rTest of ASM vs PSM.--------------------------------------------

double rTest(){
    double accumulator = 0.0,covariance = 0.0;

    for(int i = 0;i<32;i++){
       accumulator += (asM[i] - calcMean(asM)) * (psm[i] - calcMean(asM));
    }
        covariance = accumulator/31.0;

    return (covariance / (calcStdDeviation(asM,calcMean(asM)) * calcStdDeviation(psm,calcMean(psm))));
}

//--------------This function will calculate Mean of ASM & PSM.-----------------------------------------------

double calcMean(vector <double>& vec) {
    double sum = 0.0;
    for(int i = 0; i<32;i++){
        sum += vec[i];
    }
    return (sum/32.0);
}

//--------------This function will calculate Standard Deviation of ASM & PSM.---------------------------------

double calcStdDeviation(vector <double>& dev, double mean ){
    double result = 0.0;
    for(int i = 0; i<32;i++){
        result += ((dev[i] - mean) * (dev[i] - mean));
    }

    return (sqrt(result / 32.0));
}

//--------------This function will calculate Maximum of ASM & PSM.--------------------------------------------

double calcMax(vector <double>& max){
    double result = 0.0;
    for(int i=0;i<32;i++)
    {
        if(max[i]>result)
        result=max[i];
    }

    return result;
}

//--------------This function will calculate Minimum of ASM & PSM.--------------------------------------------

double calcMin(vector <double>& min){
    double result = 0.0;
    for(int i=0;i<32;i++)
    {
        if(min[i]<result)
        result=min[i];
    }

    return result;
}

//--------------This function will print all the output data in my program.-----------------------------------

void printTeamStats(){
    cout << "\n|-----------------------------------------------------------------------------------------------------------------------|";
    cout << "\n|----------TEAM NAME----------------------------------ASM------------------------------------------PSM------------------|";
    cout << "\n|-----------------------------------------------------------------------------------------------------------------------|\n";

    for(int i = 0; i<TeamName.size();i++){
        if (i == 0)
            cout <<"\nNFC East:\n";
        else if (i == 4)
            cout <<"\nNFC West:\n";
        else if (i == 8)
            cout <<"\nNFC North:\n";
        else if (i == 12)
            cout <<"\nNFC South:\n";
        else if (i == 16)
            cout <<"\nAFC East:\n";
        else if (i == 20)
            cout <<"\nAFC West:\n";
        else if (i == 24)
            cout <<"\nAFC North:\n";
        else if (i == 28)
            cout <<"\nAFC South:\n";

    cout << setw(22) << TeamName[i] << "              " << fixed << setprecision(2) << setw(22) << asM[i] << "                      " << setw(24) << psm[i] << "\n";
    }
    cout << "\n|-----------------------------------------------------------------------------------------------------------------------|\n"; 
    cout << setw(22) << "\nSample Size : " << setw(44) << 32;
    cout << setw(22) << "\nR-Test (ASM vs PSM) : " << setw(36) << setprecision(4) << rTest();
    cout << setw(22) << "\n\nMean of ASM : " << setw(44) << setprecision(4) << calcMean(asM);
    cout << setw(22) << "\nStandard Deviation of ASM : " << setw(30) << setprecision(4) << calcStdDeviation(asM,calcMean(asM));
    cout << setw(22) << "\nMax of ASM : " << setw(45) << setprecision(4) << calcMax(asM);
    cout << setw(22) << "\nMin of ASM : " << setw(45)<< setprecision(4) << calcMin(asM);
    cout << setw(22) << "\n\nMean of PSM : " << setw(44) << setprecision(4) << calcMean(psm);
    cout << setw(22) << "\nStandard Deviation of PSM : " << setw(30) << setprecision(4) << calcStdDeviation(psm,calcMean(psm));
    cout << setw(22) << "\nMax of PSM : " << setw(45) << setprecision(4) << calcMax(psm);
    cout << setw(22) << "\nMin of PSM : " << setw(45) << setprecision(4) << calcMin(psm);
}

//-----------------------END OF PROGRAM--------------------------------------------------------------------------------

File: FootballStats.java
import java.io.*;
import java.util.*;
import org.apache.commons.math3.stat.inference.*;
import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;

public class FootballStats {

    //----------------------------Declarations-------------------------------------------------------------------

    public static final int NUMBER_OF_TEAMS = 32;
    public static ArrayList<String> teams = new ArrayList<>();
    public static ArrayList<Double> asm = new ArrayList<>();
    public static ArrayList<Double> py = new ArrayList<>();
    public static ArrayList<Double> dpy = new ArrayList<>();
    public static ArrayList<Double> ry = new ArrayList<>();
    public static ArrayList<Double> dry = new ArrayList<>();
    public static ArrayList<Double> to = new ArrayList<>();
    public static ArrayList<Double> dto = new ArrayList<>();
    public static ArrayList<Double> pendif = new ArrayList<>();
    public static ArrayList<Double> rettd = new ArrayList<>();
    public static ArrayList<Double> psm = new ArrayList<>();
    public static String line = null;
    private DescriptiveStatistics statsASM = new DescriptiveStatistics();
    private DescriptiveStatistics statsPSM = new DescriptiveStatistics();
    private TTest ttest = new TTest();
    private double[] sample1 = new double[NUMBER_OF_TEAMS];
    private double[] sample2 = new double[NUMBER_OF_TEAMS];

    //---------------------------Main function-------------------------------------------------------------------

    public static void main(String [] args) {
        FootballStats test = new FootballStats();
        test.readFileAndParse();
        test.calculateTeamStats();
        test.printTeamStats();
    }

    //-------------This function will read and parse text file "teamStats.txt" into teamName & teamData----------

    public void readFileAndParse(){

        String fileName = "teamStats.txt";

        try {
            FileReader fileReader = new FileReader(fileName);
            BufferedReader bufferedReader = new BufferedReader(fileReader);

            while((line = bufferedReader.readLine()) != null) {
                String [] teamName = line.split("\\|");             //Split team name from rest data.
                String [] teamsData = teamName[1].split(" ");       //Split name of rest Data from team name.

                //Store each team name into ArrayList of "teams"
                String individualTeamName = teamName[0];
                teams.add(individualTeamName);

                //Store each data value for each team into their respective Array List. Such as asm for all teams are stored in ArrayList "asm"
                for(int i = 1; i<teamsData.length;i++){
                    double aDouble = Double.parseDouble(teamsData[i]);    //Convert string into double so we can apply Mathematic calculation on it.

                    if(i == 1)
                        asm.add(aDouble);
                    else if(i == 2)
                        py.add(aDouble);
                    else if(i == 3)
                        dpy.add(aDouble);
                    else if(i== 4)
                        ry.add(aDouble);
                    else if(i== 5)
                        dry.add(aDouble);
                    else if(i== 6)
                        to.add(aDouble);
                    else if(i== 7)
                        dto.add(aDouble);
                    else if(i== 8)
                        pendif.add(aDouble);
                    else if(i== 9)
                        rettd.add(aDouble);
                    else
                        System.out.println("Error in Storing Data ! \n");
                }
            }
            bufferedReader.close();
        }

        catch(FileNotFoundException ex) {
            System.out.println(
                    "Unable to open file '" +        //give error if unable to open given file.
                            fileName + "'");
        }
        catch(IOException ex) {
            System.out.println(
                    "Error reading file '"           //give error if there is problem in reading given file.
                            + fileName + "'");
        }
    }

    //-------------This function will calculate PSM and store ASM PSM into sample1[] sample2[] for tTest.

    public void calculateTeamStats() {

        for (int i = 0; i < NUMBER_OF_TEAMS; i++) {
            double ans = 3.17 * (rettd.get(i)) - 0.06 * (pendif.get(i)) + 61.67 * (py.get(i)) + 26.44 * (ry.get(i)) - 2.77 * (to.get(i)) - 67.5 * (dpy.get(i)) - 22.79 * (dry.get(i)) + 3.49 * (dto.get(i));
            psm.add(ans);
            sample1[i] = asm.get(i);
            sample2[i] = psm.get(i);
            statsASM.addValue(sample1[i]);
            statsPSM.addValue(sample2[i]);
        }
    }

    //--------------This function will print all the output data in my program.-----------------------------------

    public void printTeamStats() {

        System.out.println("|-----------------------------------------------------------------------------------------------------------------------|");
        System.out.println("|----------TEAM NAME----------------------------------ASM------------------------------------------PSM------------------|");
        System.out.println("|-----------------------------------------------------------------------------------------------------------------------|");

        for (int i = 0; i < NUMBER_OF_TEAMS; i++) {
            if (i == 0)
                System.out.println("\nNFC East:");
            else if (i == 4)
                System.out.println("\nNFC West:");
            else if (i == 8)
                System.out.println("\nNFC North:");
            else if (i == 12)
                System.out.println("\nNFC South:");
            else if (i == 16)
                System.out.println("\nAFC East:");
            else if (i == 20)
                System.out.println("\nAFC West:");
            else if (i == 24)
                System.out.println("\nAFC North:");
            else if (i == 28)
                System.out.println("\nAFC South:");

            System.out.printf("%22s", teams.get(i));
            System.out.printf("%36.2f", asm.get(i));
            System.out.printf("%45.2f \n", psm.get(i));
        }
        System.out.println("|-----------------------------------------------------------------------------------------------------------------------|");
        System.out.printf("\nSample Size : %25s", NUMBER_OF_TEAMS);
        System.out.printf("\nT-Test (ASM vs PSM) : %17.4f ", ttest.t(sample1, sample2));
        System.out.printf("\n\nMean of ASM : %25.4f ", statsASM.getMean());
        System.out.printf("\nStandard Deviation of ASM : %11.4f ", statsASM.getStandardDeviation());
        System.out.printf("\nMax of ASM : %26.4f ", statsASM.getMax());
        System.out.printf("\nMin of ASM : %26.4f ", statsASM.getMin());
        System.out.printf("\n\nMean of PSM : %25.4f ", statsPSM.getMean());
        System.out.printf("\nStandard Deviation of PSM : %11.4f ", statsPSM.getStandardDeviation());
        System.out.printf("\nMax of PSM : %26.4f ", statsPSM.getMax());
        System.out.printf("\nMin of PSM : %26.4f ", statsPSM.getMin());
    }
}

//-----------------------END OF PROGRAM--------------------------------------------------------------------------------